lib/nghttp2_hd.c

```diff
@@ -594,8 +594,20 @@ static void hd_map_remove(nghttp2_hd_map *map, nghttp2_hd_entry *ent) {
 static int hd_ringbuf_init(nghttp2_hd_ringbuf *ringbuf, size_t bufsize,
                            nghttp2_mem *mem) {
   size_t size;
-  for (size = 1; size < bufsize; size <<= 1)
+  size_t max_size;
+
+  max_size = SIZE_MAX / sizeof(nghttp2_hd_entry *);
+  if (bufsize > max_size) {
+    return NGHTTP2_ERR_NOMEM;
+  }
+
+  for (size = 1; size < bufsize && size < max_size / 2; size <<= 1)
     ;
+
+  if (size * sizeof(nghttp2_hd_entry *) > max_size) {
```

Comment:
```suggestion
  if (size > max_size) {
```

Reply:
true, `max_size` already accounts for `sizeof(nghttp2_hd_entry *)`